## User-User Collaborative Filtering Assignment

In [3]:
import pandas as pd
import numpy as np

### Part 1 - Without Normalization

#### Import the data

In [4]:
uu_df = pd.read_excel('./UUCF Assignment Spreadsheet.xls',sheetname=0)

In [5]:
uu_df.shape

(100, 25)

In [6]:
uu_df

1648  5136  918   2824  \
11: Star Wars: Episode IV - A New Hope (1977)        NaN   4.5   5.0   4.5   
12: Finding Nemo (2003)                              NaN   5.0   5.0   NaN   
13: Forrest Gump (1994)                              NaN   5.0   4.5   5.0   
14: American Beauty (1999)                           NaN   4.0   NaN   NaN   
22: Pirates of the Caribbean: The Curse of the ...   4.0   5.0   3.0   4.5   
24: Kill Bill: Vol. 1 (2003)                         3.0   5.0   NaN   4.0   
38: Eternal Sunshine of the Spotless Mind (2004)     NaN   5.0   5.0   NaN   
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)        NaN   3.0   NaN   NaN   
77: Memento (2000)                                   NaN   NaN   5.0   5.0   
85: Raiders of the Lost Ark (Indiana Jones and ...   NaN   5.0   NaN   NaN   
98: Gladiator (2000)                                 NaN   3.5   NaN   5.0   
105: Back to the Future (1985)                       4.5   5.0   5.0   NaN   
107: Snatch (2000)                                   NaN   NaN   NaN   5.0   
114: Pretty Woman (1990)                             NaN   3.5   NaN   NaN   
120: The Lord of the Rings: The Fellowship of t...   NaN   5.0   NaN   4.0   
121: The Lord of the Rings: The Two Towers (2002)    NaN   5.0   4.5   4.0   
122: The Lord of the Rings: The Return of the K...   NaN   5.0   NaN   4.0   
134: O Brother Where Art Thou? (2000)                NaN   NaN   NaN   NaN   
141: Donnie Darko (2001)                             NaN   3.0   NaN   NaN   
146: Crouching Tiger Hidden Dragon (Wo hu cang ...   2.5   4.0   NaN   NaN   
153: Lost in Translation (2003)                      NaN   4.5   NaN   NaN   
155: The Dark Knight (2008)                          5.0   5.0   NaN   5.0   
161: Ocean's Eleven (2001)                           4.0   4.5   NaN   NaN   
180: Minority Report (2002)                          3.0   4.5   NaN   4.5   
187: Sin City (2005)                                 NaN   3.0   4.0   NaN   
194: Amelie (2001)                                   NaN   5.0   NaN   NaN   
197: Braveheart (1995)                               NaN   4.0   NaN   NaN   
238: The Godfather (1972)                            NaN   5.0   NaN   NaN   
243: High Fidelity (2000)                            NaN   4.0   NaN   NaN   
268: Batman (1989)                                   NaN   NaN   NaN   NaN   
...                                                  ...   ...   ...   ...   
862: Toy Story (1995)                                4.0   5.0   5.0   NaN   
954: Mission: Impossible (1996)                      4.5   4.0   4.0   NaN   
955: Mission: Impossible II (2000)                   4.0   3.5   4.0   NaN   
1422: The Departed (2006)                            NaN   5.0   NaN   3.5   
1572: Die Hard: With a Vengeance (1995)              NaN   NaN   NaN   NaN   
1597: Meet the Parents (2000)                        NaN   2.0   NaN   NaN   
1637: Speed (1994)                                   NaN   3.0   NaN   NaN   
1891: Star Wars: Episode V - The Empire Strikes...   NaN   4.5   5.0   4.0   
1892: Star Wars: Episode VI - Return of the Jed...   NaN   4.5   5.0   4.5   
1894: Star Wars: Episode II - Attack of the Clo...   3.0   2.5   5.0   4.0   
1900: Traffic (2000)                                 NaN   1.0   NaN   NaN   
2024: The Patriot (2000)                             NaN   NaN   NaN   NaN   
2164: Stargate (1994)                                NaN   NaN   NaN   NaN   
2501: The Bourne Identity (2002)                     NaN   4.5   NaN   NaN   
2502: The Bourne Supremacy (2004)                    NaN   4.0   NaN   NaN   
3049: Ace Ventura: Pet Detective (1994)              NaN   3.0   4.0   NaN   
4327: Charlie's Angels (2000)                        NaN   1.5   NaN   NaN   
5503: The Fugitive (1993)                            NaN   4.0   NaN   NaN   
7443: Chicken Run (2000)                             NaN   1.0   NaN   NaN   
8358: Cast Away (2000)                               NaN   4.0   NaN   NaN

#### Create the user-user correlation matrix

In [7]:
uu_corr_df = uu_df.corr()

In [8]:
uu_corr_df.head()

1648      5136      918       2824      3867      860       3712  \
1648  1.000000  0.402980 -0.142206  0.517620  0.300200  0.480537 -0.312412   
5136  0.402980  1.000000  0.118979  0.057916  0.341734  0.241377  0.131398   
918  -0.142206  0.118979  1.000000 -0.317063  0.294558  0.468333  0.092037   
2824  0.517620  0.057916 -0.317063  1.000000 -0.060913 -0.008066  0.462910   
3867  0.300200  0.341734  0.294558 -0.060913  1.000000  0.282497  0.400275   

          2968      3525      4323    ...         3556      5261      2492  \
1648  0.383348  0.092775  0.098191    ...    -0.191988  0.493008  0.360644   
5136  0.206695  0.360056  0.033642    ...     0.488607  0.328120  0.422236   
918  -0.045854  0.367568 -0.035394    ...     0.373226  0.470972  0.069956   
2824  0.214760  0.169907  0.119350    ...    -0.201275  0.228341  0.238700   
3867  0.264249  0.125193 -0.333602    ...     0.174085  0.297977  0.476683   

          5062      2486      4942      2267      4809      3853      2288  
1648  0.551089  0.002544  0.116653 -0.429183  0.394371 -0.304422  0.245048  
5136  0.226635  0.305803  0.037769  0.240728  0.411676  0.189234  0.390067  
918  -0.054762  0.133812  0.015169 -0.273096  0.082528  0.667168  0.119162  
2824  0.259660  0.247097  0.149247 -0.361466  0.474974 -0.262073  0.166999  
3867  0.293868  0.438992 -0.162818 -0.295966  0.054518  0.464110  0.379856  

[5 rows x 25 columns]

Note that the correlation between users 1648 and 5136 is 0.40298, and the correlation between users 918 and 2824 is -0.31706.

In [9]:
uu_corr_df.loc[[1648,918],[5136,2824]]

5136      2824
1648  0.402980  0.517620
918   0.118979 -0.317063

#### function to indentify the top-n neighbors

In [10]:
uu_corr_df.loc[1648].sort_values(ascending=False ).iloc[1:6]

5062    0.551089
2824    0.517620
5261    0.493008
860     0.480537
89      0.464610
Name: 1648, dtype: float64

In [11]:
def top_n_neighbors(user_movie_df,user,n=5):
    uu_corr_df = user_movie_df.corr()
    return uu_corr_df.loc[user].sort_values(ascending=False ).iloc[1:n+1] 

In [12]:
top_n_neighbors(uu_df,3867,5)

2492    0.476683
3853    0.464110
2486    0.438992
3712    0.400275
2288    0.379856
Name: 3867, dtype: float64

In [13]:
top_n_neighbors(uu_df,1648,5)

5062    0.551089
2824    0.517620
5261    0.493008
860     0.480537
89      0.464610
Name: 1648, dtype: float64

Test: if the target user were #3712, the closest neighbors are 2824 (corr: 0.46291), 3867 (corr: 0.400275), 5062 (corr: 0.247693), 442 (corr: 0.22713), and 3853 (corr: 0.19366)

In [14]:
top_nn_df = top_n_neighbors(uu_df,3712,5)

In [15]:
print(top_nn_df)

2824    0.462910
3867    0.400275
5062    0.247693
442     0.227130
3853    0.193660
Name: 3712, dtype: float64


Compute the predictions for each movie by taking the correlation-weighted average of the ratings of the top-five neighbors (for each target user) for each movie.

In [16]:
ratings_top_nn_df = uu_df.loc[:,top_nn_df.index]

In [17]:
ratings_top_nn_df.head()

2824  3867  5062  442   \
11: Star Wars: Episode IV - A New Hope (1977)        4.5   4.0   4.0   3.0   
12: Finding Nemo (2003)                              NaN   4.0   4.0   NaN   
13: Forrest Gump (1994)                              5.0   4.5   4.5   3.0   
14: American Beauty (1999)                           NaN   NaN   4.5   4.0   
22: Pirates of the Caribbean: The Curse of the ...   4.5   4.0   4.0   1.0   

                                                    3853  
11: Star Wars: Episode IV - A New Hope (1977)        NaN  
12: Finding Nemo (2003)                              NaN  
13: Forrest Gump (1994)                              3.5  
14: American Beauty (1999)                           NaN  
22: Pirates of the Caribbean: The Curse of the ...   NaN

In [18]:
ratings_top_nn_df.notnull()

2824   3867   5062  \
11: Star Wars: Episode IV - A New Hope (1977)        True   True   True   
12: Finding Nemo (2003)                             False   True   True   
13: Forrest Gump (1994)                              True   True   True   
14: American Beauty (1999)                          False  False   True   
22: Pirates of the Caribbean: The Curse of the ...   True   True   True   
24: Kill Bill: Vol. 1 (2003)                         True   True   True   
38: Eternal Sunshine of the Spotless Mind (2004)    False  False   True   
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)       False  False  False   
77: Memento (2000)                                   True  False   True   
85: Raiders of the Lost Ark (Indiana Jones and ...  False   True   True   
98: Gladiator (2000)                                 True   True   True   
105: Back to the Future (1985)                      False  False   True   
107: Snatch (2000)                                   True  False   True   
114: Pretty Woman (1990)                            False   True  False   
120: The Lord of the Rings: The Fellowship of t...   True   True   True   
121: The Lord of the Rings: The Two Towers (2002)    True   True   True   
122: The Lord of the Rings: The Return of the K...   True   True   True   
134: O Brother Where Art Thou? (2000)               False  False   True   
141: Donnie Darko (2001)                            False  False   True   
146: Crouching Tiger Hidden Dragon (Wo hu cang ...  False  False   True   
153: Lost in Translation (2003)                     False  False   True   
155: The Dark Knight (2008)                          True   True   True   
161: Ocean's Eleven (2001)                          False   True   True   
180: Minority Report (2002)                          True   True   True   
187: Sin City (2005)                                False  False   True   
194: Amelie (2001)                                  False  False   True   
197: Braveheart (1995)                              False  False  False   
238: The Godfather (1972)                           False   True  False   
243: High Fidelity (2000)                           False  False   True   
268: Batman (1989)                                  False   True  False   
...                                                   ...    ...    ...   
862: Toy Story (1995)                               False   True   True   
954: Mission: Impossible (1996)                     False  False   True   
955: Mission: Impossible II (2000)                  False  False   True   
1422: The Departed (2006)                            True   True  False   
1572: Die Hard: With a Vengeance (1995)             False   True   True   
1597: Meet the Parents (2000)                       False   True   True   
1637: Speed (1994)                                  False   True   True   
1891: Star Wars: Episode V - The Empire Strikes...   True   True   True   
1892: Star Wars: Episode VI - Return of the Jed...   True   True   True   
1894: Star Wars: Episode II - Attack of the Clo...   True   True   True   
1900: Traffic (2000)                                False  False  False   
2024: The Patriot (2000)                            False  False  False   
2164: Stargate (1994)                               False  False  False   
2501: The Bourne Identity (2002)                    False   True   True   
2502: The Bourne Supremacy (2004)                   False   True   True   
3049: Ace Ventura: Pet Detective (1994)             False   True  False   
4327: Charlie's Angels (2000)                       False   True   True   
5503: The Fugitive (1993)                           False  False  False   
7443: Chicken Run (2000)                            False  False   True   
8358: Cast Away (2000)                              False  False  False   
8467: Dumb & Dumber (1994)                          False   True   True   
8587: The Lion King (1994)                           True   True  False   
933

In [19]:
den_pred_df = (ratings_top_nn_df.notnull()).mul(top_nn_df).sum(axis=1) # denominator prediction formula

In [20]:
den_pred_df

11: Star Wars: Episode IV - A New Hope (1977)                                         1.338008
12: Finding Nemo (2003)                                                               0.647968
13: Forrest Gump (1994)                                                               1.531667
14: American Beauty (1999)                                                            0.474823
22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)                     1.338008
24: Kill Bill: Vol. 1 (2003)                                                          1.531667
38: Eternal Sunshine of the Spotless Mind (2004)                                      0.668483
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)                                         0.420789
77: Memento (2000)                                                                    1.131393
85: Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    1.068757
98: Gladiator (2000)                              

In [21]:
ratings_top_nn_df = uu_df.loc[:,top_nn_df.index].fillna(0) #movie ratings from top n users

In [22]:
print(ratings_top_nn_df)

                                                    2824  3867  5062  442   \
11: Star Wars: Episode IV - A New Hope (1977)        4.5   4.0   4.0   3.0   
12: Finding Nemo (2003)                              0.0   4.0   4.0   0.0   
13: Forrest Gump (1994)                              5.0   4.5   4.5   3.0   
14: American Beauty (1999)                           0.0   0.0   4.5   4.0   
22: Pirates of the Caribbean: The Curse of the ...   4.5   4.0   4.0   1.0   
24: Kill Bill: Vol. 1 (2003)                         4.0   3.0   4.0   5.0   
38: Eternal Sunshine of the Spotless Mind (2004)     0.0   0.0   4.0   3.0   
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)        0.0   0.0   0.0   2.0   
77: Memento (2000)                                   5.0   0.0   4.0   4.0   
85: Raiders of the Lost Ark (Indiana Jones and ...   0.0   4.5   4.0   5.0   
98: Gladiator (2000)                                 5.0   4.0   4.0   4.0   
105: Back to the Future (1985)                       0.0   0.0  

In [23]:
num_pred_df = ratings_top_nn_df.mul(top_nn_df).sum(axis=1) # numerator of prediction formula

In [24]:
num_pred_df.head()

11: Star Wars: Episode IV - A New Hope (1977)                        5.356356
12: Finding Nemo (2003)                                              2.591871
13: Forrest Gump (1994)                                              6.589603
14: American Beauty (1999)                                           2.023139
22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)    4.902096
dtype: float64

In [25]:
print(num_pred_df.loc['603: The Matrix (1999)'])

7.43766002895


In [26]:
print(num_pred_df.loc['105: Back to the Future (1985)'])

2.25026904998


In [27]:
pre_movies_df = num_pred_df.div(den_pred_df,axis='index')

In [28]:
pre_movies_df

11: Star Wars: Episode IV - A New Hope (1977)                                         4.003233
12: Finding Nemo (2003)                                                               4.000000
13: Forrest Gump (1994)                                                               4.302242
14: American Beauty (1999)                                                            4.260827
22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)                     3.663728
24: Kill Bill: Vol. 1 (2003)                                                          4.013394
38: Eternal Sunshine of the Spotless Mind (2004)                                      3.949931
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)                                         2.920458
77: Memento (2000)                                                                    4.494735
85: Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    4.399779
98: Gladiator (2000)                              

In [29]:
print(pre_movies_df.mask(np.isinf(pre_movies_df)).dropna().sort_values(ascending=False ).iloc[0:12] )

641: Requiem for a Dream (2000)                                                       5.000000
603: The Matrix (1999)                                                                4.855924
105: Back to the Future (1985)                                                        4.739173
107: Snatch (2000)                                                                    4.651432
155: The Dark Knight (2008)                                                           4.622564
121: The Lord of the Rings: The Two Towers (2002)                                     4.616916
120: The Lord of the Rings: The Fellowship of the Ring (2001)                         4.616916
122: The Lord of the Rings: The Return of the King (2003)                             4.571337
77: Memento (2000)                                                                    4.494735
1572: Die Hard: With a Vengeance (1995)                                               4.457406
85: Raiders of the Lost Ark (Indiana Jones and the

In [37]:
def predict_item(user_movie_df,user,n=5,l=12):
    top_nn_df = top_n_neighbors(user_movie_df,user,n)#top n correlated users
    ratings_top_nn_df = uu_df.loc[:,top_nn_df.index] #movie ratings from top n users
    den_pred_df = ratings_top_nn_df.notnull().mul(top_nn_df).sum(axis=1) # denominator prediction formula
    ratings_top_nn_df.fillna(0) #movie ratings from top n users
    num_pred_df = ratings_top_nn_df.mul(top_nn_df).sum(axis=1) # numerator of prediction formula
    pred_items_df = num_pred_df.div(den_pred_df)
    return  pred_items_df.sort_values(ascending=False ).iloc[0:l]

Test:  if the user ID was 3712, the correct submission would be:

Top Movie: 641 Prediction: 5.000

2nd Movie: 603 Prediction: 4.856

3rd Movie: 105 Prediction: 4.739

And if the user ID were 3525, 

there would be a three-way tie among:

Movies: 238, 194, and 38 (all with a prediction of 5.000)

In [38]:
predict_item(uu_df,3867,5,3)

1891: Star Wars: Episode V - The Empire Strikes Back (1980)    4.760291
155: The Dark Knight (2008)                                    4.551454
122: The Lord of the Rings: The Return of the King (2003)      4.507637
dtype: float64

In [39]:
predict_item(uu_df,89,5,3)

238: The Godfather (1972)               4.894124
278: The Shawshank Redemption (1994)    4.882194
807: Seven (a.k.a. Se7en) (1995)        4.774093
dtype: float64

### Part 2 - Normalization

In [47]:
ratings_top_nn_df = uu_df.loc[:,top_nn_df.index]

In [48]:
mean_user =  uu_df[3712].mean()

In [49]:
print(mean_user)

4.5


In [50]:
mean_top_nn_df = ratings_top_nn_df.mean(axis=0)

In [51]:
mean_top_nn_df

2824    4.058824
3867    3.661538
5062    3.865672
442     3.600000
3853    3.700000
dtype: float64

In [55]:
ratings_top_nn_df-mean_top_nn_df

2824      3867  \
11: Star Wars: Episode IV - A New Hope (1977)       0.441176  0.338462   
12: Finding Nemo (2003)                                  NaN  0.338462   
13: Forrest Gump (1994)                             0.941176  0.838462   
14: American Beauty (1999)                               NaN       NaN   
22: Pirates of the Caribbean: The Curse of the ...  0.441176  0.338462   
24: Kill Bill: Vol. 1 (2003)                       -0.058824 -0.661538   
38: Eternal Sunshine of the Spotless Mind (2004)         NaN       NaN   
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)            NaN       NaN   
77: Memento (2000)                                  0.941176       NaN   
85: Raiders of the Lost Ark (Indiana Jones and ...       NaN  0.838462   
98: Gladiator (2000)                                0.941176  0.338462   
105: Back to the Future (1985)                           NaN       NaN   
107: Snatch (2000)                                  0.941176       NaN   
114: Pretty Woman (1990)                                 NaN  0.338462   
120: The Lord of the Rings: The Fellowship of t... -0.058824  1.338462   
121: The Lord of the Rings: The Two Towers (2002)  -0.058824  1.338462   
122: The Lord of the Rings: The Return of the K... -0.058824  1.338462   
134: O Brother Where Art Thou? (2000)                    NaN       NaN   
141: Donnie Darko (2001)                                 NaN       NaN   
146: Crouching Tiger Hidden Dragon (Wo hu cang ...       NaN       NaN   
153: Lost in Translation (2003)                          NaN       NaN   
155: The Dark Knight (2008)                         0.941176  1.338462   
161: Ocean's Eleven (2001)                               NaN  0.338462   
180: Minority Report (2002)                         0.441176 -0.161538   
187: Sin City (2005)                                     NaN       NaN   
194: Amelie (2001)                                       NaN       NaN   
197: Braveheart (1995)                                   NaN       NaN   
238: The Godfather (1972)                                NaN -0.661538   
243: High Fidelity (2000)                                NaN       NaN   
268: Batman (1989)                                       NaN -1.161538   
...                                                      ...       ...   
862: Toy Story (1995)                                    NaN -0.661538   
954: Mission: Impossible (1996)                          NaN       NaN   
955: Mission: Impossible II (2000)                       NaN       NaN   
1422: The Departed (2006)                          -0.558824 -0.661538   
1572: Die Hard: With a Vengeance (1995)                  NaN  1.338462   
1597: Meet the Parents (2000)                            NaN -0.161538   
1637: Speed (1994)                                       NaN -1.661538   
1891: Star Wars: Episode V - The Empire Strikes... -0.058824  0.838462   
1892: Star Wars: Episode VI - Return of the Jed...  0.441176  0.338462   
1894: Star Wars: Episode II - Attack of the Clo... -0.058824 -0.161538   
1900: Traffic (2000)                                     NaN       NaN   
2024: The Patriot (2000)                                 NaN       NaN   
2164: Stargate (1994)                                    NaN       NaN   
2501: The Bourne Identity (2002)                         NaN -0.161538   
2502: The Bourne Supremacy (2004)                        NaN  0.338462   
3049: Ace Ventura: Pet Detective (1994)                  NaN -1.161538   
4327: Charlie's Angels (2000)                            NaN -0.661538   
5503: The Fugitive (1993)                                NaN       NaN   
7443: Chicken Run (2000)                                 NaN       NaN   
8358: Cast Away (2000)                                   NaN       NaN   
8467: Dumb & Dumber (1994)                               NaN -2.661538   
8587: The Lion King (1994)                         -0.558824  0.338462   
9331: Clear and Present Danger (1994)                    NaN

In [56]:
def predict_item_normalized(user_movie_df,user,n=5,l=12):
    top_nn_df = top_n_neighbors(user_movie_df,user,n)#top n correlated users
    ratings_top_nn_df = uu_df.loc[:,top_nn_df.index] #movie ratings from top n users
    
    mean_user =  user_movie_df[user].mean()
    
    den_pred_df = ratings_top_nn_df.notnull().mul(top_nn_df).sum(axis=1) # denominator prediction formula
    
    mean_top_nn_df = ratings_top_nn_df.mean(axis=0)
    
    ratings_top_nn_df.fillna(0) #movie ratings from top n users
    num_pred_df = (ratings_top_nn_df-mean_top_nn_df).mul(top_nn_df).sum(axis=1) # numerator of prediction formula
    pred_items_df = mean_user +  num_pred_df.div(den_pred_df)
    return  pred_items_df.sort_values(ascending=False ).iloc[0:l]

For example, if the user ID was 3712, the correct submission would be:

Top Movie: 641 Prediction: 5.900

2nd Movie: 603 Prediction: 5.546

3rd Movie: 105 Prediction: 5.501

In [58]:
predict_item_normalized(uu_df,3712,5,3)

641: Requiem for a Dream (2000)    5.900000
603: The Matrix (1999)             5.545567
105: Back to the Future (1985)     5.500585
dtype: float64

In [59]:
predict_item_normalized(uu_df,3867,5,3)

1891: Star Wars: Episode V - The Empire Strikes Back (1980)    5.245509
155: The Dark Knight (2008)                                    4.856770
77: Memento (2000)                                             4.777803
dtype: float64

In [60]:
predict_item_normalized(uu_df,89,5,3)

238: The Godfather (1972)               5.322015
278: The Shawshank Redemption (1994)    5.261424
275: Fargo (1996)                       5.241111
dtype: float64